### PR TITLE
hierarchies: Change `HierarchyProvider.getNodes` from `Promise<HierarchyNode[]>` to `AsyncIterableIterator<HierarchyNode>`

### PR DESCRIPTION
**Pull request by grigasp**
Stay consistent with `getNodeInstanceKeys` and provides ability to stream nodes as they get loaded rather than having to wait for the whole list.

### CODE REVIEW

**Comment by Yato333** [comment]
The performance of large flat hierarchy and base class grouping has worsened dramatically. Why would that be?

File: packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+import { Observable } from "rxjs";
+
+/**
+ * This is pretty much a copy of https://github.com/benlesh/rxjs-for-await/blob/94f9cf9cb015ac3700dfd1850eb81d36962eb70f/src/index.ts#L33,
+ * except that we're using a linked list-based queue rather than an array-based one.
+ *
+ * @internal
+ */
+export async function* eachValueFrom<T>(source: Observable<T>): AsyncIterableIterator<T> {
+  const deferreds = new Queue<ResolvablePromise<{ value?: T; done: boolean }>>();
```
Comment (Yato333) [inline packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts:15]: Is there a point for copying the exact same code? 
`deferred` variable can be renamed to something like `deferredPromises`, `null` can be changed to `undefined`, and all `else`s in `while(true)` can be removed.

**Comment by grigasp** [comment]
> The performance of large flat hierarchy and base class grouping has worsened dramatically. Why would that be?

Turns out that was due to excessive use of `Array.shift` in `eachValueFrom` implementation that we were using. Changed the implementation to use linked list instead - the performance is back on track now.

File: packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts (hunk 1 above)
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts:15]: don't see a point in renaming `deferred` to `deferredPromise`, but agree about the rest - cleaned up

File: packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts
Comment (Yato333) [inline packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts:1]: This isn't an operator. Shouldn't it go to the internal folder?

File: packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts (hunk 1 above)
Comment (Yato333) [reply packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts:15]: Sorry, I meant `deferreds` variable which is not a word and was added to the spell checker exceptions

File: packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts:1]: Monday... :D

File: packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts (hunk 1 above)
Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts]: I think that's valid word: https://www.dictionary.com/browse/deferred

Comment (Yato333) [reply packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts]: It says that deferred is an adjective and here it's used as a noun in plural. I'm not sure if it's valid

Comment (grigasp) [reply packages/hierarchies/src/hierarchies/internal/operators/EachValueFrom.ts]: let's try this, then :) https://en.wiktionary.org/wiki/deferreds#English

**Review by Yato333 — APPROVED** [review]

**Pull request merged**